### PR TITLE
add capability of handling null in string related built-in functions

### PR DESCRIPTION
**Pull request by cosmicshuai**
close: https://github.com/microsoft/botbuilder-dotnet/issues/2709

related modified functions are:

- length
- replace
- repalceIgnoreCase
- split
- substring
- toLower
- toUpper
- trim
- endsWith
- startsWith
- countWord
- concat
- indexOf
- lastIndexOf

### CODE REVIEW

File: libraries/Microsoft.Bot.Expressions/BuiltInFunctions.cs [hunk 1]
```diff
@@ -377,6 +377,17 @@ public static string VerifyString(object value, Expression expression, int numbe
             return error;
         }
 
+        public static string VerifyStringOrNull(object value, Expression expression, int number)
+        {
+            string error = null;
+            if (!(value is string) && !(value == null))
```
Comment (feich-ms) [inline libraries/Microsoft.Bot.Expressions/BuiltInFunctions.cs]: !(value == null) can be value != null

File: libraries/Microsoft.Bot.Expressions/BuiltInFunctions.cs [hunk 2]
```diff
@@ -2568,56 +2583,269 @@ private static EvaluateExpressionDelegate SortBy(bool isDescending)
                 // String
                 new ExpressionEvaluator(
                     ExpressionType.Concat,
-                    Apply(
+                    ApplyWithError(
                         args =>
                         {
                             var builder = new StringBuilder();
+                            string error = null;
```
Comment (feich-ms) [inline libraries/Microsoft.Bot.Expressions/BuiltInFunctions.cs]: error is always null, so why use ApplyWithError?

File: libraries/Microsoft.Bot.Expressions/BuiltInFunctions.cs [hunk 3]
```diff
@@ -2568,56 +2583,269 @@ private static EvaluateExpressionDelegate SortBy(bool isDescending)
...
+                                }),
                 new ExpressionEvaluator(
                     ExpressionType.StartsWith,
-                    Apply(args => args[0].StartsWith(args[1]), VerifyString),
+                    Apply(
+                        args =>
+                        {
+                            string rawStr = null;
+                            string seekStr = null;
+                            if (args[0] is string str)
```
Comment (Danieladu) [inline libraries/Microsoft.Bot.Expressions/BuiltInFunctions.cs]: extract 
```
if(s is string str)
{
                                rawStr = str;
                            }
                            else
                            {
                                rawStr = string.Empty;
                            }
```

**Comment by fuselabs** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.Luis.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.Luis/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.QnA.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.QnA/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.ApplicationInsights.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.ApplicationInsights/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Azure.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Dialogs.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Dialogs/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.TemplateManager.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.TemplateManager/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Configuration.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Configuration/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Connector.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Connector/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Schema.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Schema/4.5.3)

**Comment by fuselabs** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.Luis.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.Luis/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.QnA.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.QnA/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.ApplicationInsights.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.ApplicationInsights/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Azure.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Dialogs.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Dialogs/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.TemplateManager.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.TemplateManager/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Configuration.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Configuration/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Connector.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Connector/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Schema.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Schema/4.5.3)

File: libraries/Microsoft.Bot.Expressions/BuiltInFunctions.cs [hunk 4]
```diff
@@ -442,6 +442,21 @@ public static string VerifyBoolean(object value, Expression expression, int numb
             return error;
         }
 
+        public static string ParseStringOrNull(object value)
```
Comment (Danieladu) [inline libraries/Microsoft.Bot.Expressions/BuiltInFunctions.cs]: Why make this public?

**Comment by fuselabs** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.Luis.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.Luis/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.QnA.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.QnA/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.ApplicationInsights.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.ApplicationInsights/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Azure.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Dialogs.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Dialogs/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.TemplateManager.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.TemplateManager/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Configuration.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Configuration/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Connector.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Connector/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Schema.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Schema/4.5.3)

**Pull request merged**